Bhagabat Prasad Behera

Batch 6



# 2A Assignment

## Backpropagation Step by Step

![](https://i.ibb.co/pdZNB42/backprop22.jpg)

The primary objective of training a neural network is to find the values of its parameters such that the error is minimum. For that, there are [a few techniques](https://stats.stackexchange.com/questions/235862/is-it-possible-to-train-a-neural-network-without-backpropagation) and Backpropagration is the most popular one. In this blog, we will consider a small and simple network for better understanding how backpropagation works step by step.

### Overview and Architecture

We will build a network with three layers:
* **Input** layer with two input neurons
* **Hidden** layer with two neurons
* **Output** layer with one neuron

![](https://i.ibb.co/TkqFvgL/backprop23.jpg)

### Weight Initialization

Since the goal is to minimize the the prediction error, we will initialize the model weights to [small random positive values near to 0 and less than 1](http://ufldl.stanford.edu/tutorial/supervised/MultiLayerNeuralNetworks/). Then the backpropagation technique will use gradient descent optimization to update the weights in an attempt to correctly map arbitrary inputs to outputs.

Our initial weights will be as : `w1 = 0.22`, `w2 = 0.09`, `w3 = 0.05`, `w4 = 0.11`, `w5 = 0.18` and `w6 = 0.13`. The bias is initialized with zero.

![](https://i.ibb.co/qszJsHR/backprop8.jpg)

### Dataset

Let us take two input values and one output value. The values are as following:

inputs = [2, 3]

output = [1].

![](https://i.ibb.co/G5Z3Pn8/dataset.jpg)

### Forward Pass

We will use given weights and inputs to predict the output. Inputs are multiplied by weights; the results are then passed forward to next layer.

![](https://i.ibb.co/wYBC1LN/backprop9.jpg)

$\begin{bmatrix}2 \ 3\end{bmatrix} $. $\begin{bmatrix}0.22 & 0.05 \\ 0.09 & 0.11\end{bmatrix}$ = $\begin{bmatrix}0.71\ 0.43\end{bmatrix} $

$\begin{bmatrix}0.71\ 0.43\end{bmatrix} $.$\begin{bmatrix}0.18\\ 0.13\end{bmatrix} $ = [0.184]

2 x 0.22 + 3 x 0.9 = 0.71

2 x 0.05 + 3 x 0.11 = 0.43

0.71 x 0.18 + 0.43 x 0.13 = 0.184

### Calculating Error

Now, it’s time to find out how our network performed by calculating the difference between the actual output and predicted one. It’s clear that our network output, or prediction, is not even close to actual output. We can calculate the difference or the error as following.

![](https://i.ibb.co/xXhGRWb/backprop10.jpg)

### Reducing Error

Our main goal of the training is to reduce the error or the difference between prediction and actual output. Since actual output is constant, “not changing”, and we cannot change the input values, the only way to reduce the error is to change prediction value. The question now is, how to change prediction value?

By decomposing prediction into its basic elements we can find that weights are the variable elements in the network affecting prediction value. In other words, in order to change prediction value, we need to change weights values.

![](https://i.ibb.co/gg5PVQ7/backprop11.jpg)

The question now is **how to change\update the weights value so that the error is reduced?**
The answer is **Backpropagation**!

### Backpropagation

**Backpropagation**, short for “backward propagation of errors”, is a mechanism used to update the weights using gradient descent. It calculates the gradient of the error function with respect to the neural network’s weights. The calculation proceeds backwards through the network.

**Gradient descent** is an iterative optimization algorithm for finding the minimum of a function; in our case we want to minimize th error function. To find a local minimum of a function using gradient descent, one takes steps proportional to the negative of the gradient of the function at the current point.

![](https://i.ibb.co/LxXx1KJ/backprop12.jpg)

For example, to update `w5`, we take the current `w5` and subtract the partial derivative of error function with respect to `w5`. Optionally, we multiply the derivative of the error function by a selected number to make sure that the new updated weight is minimizing the error function; this number is called learning rate.

![](https://i.ibb.co/cXSTVDm/backprop13.jpg)

The derivation of the error function is evaluated by applying the chain rule as following

![](https://i.ibb.co/pLyMY3J/backprop14.jpg)

So to update `w5` we can apply the following formula:

![](https://i.ibb.co/HHWmZPS/backprop15.jpg)

Similarly, we can derive the update formula for w6 and any other weights existing between the output and the hidden layer.

![](https://i.ibb.co/gwLfhDh/backprop16.jpg)



However, when moving backward to update `w1`, `w2`, `w3` and `w4` existing between input and hidden layer, the partial derivative for the error function with respect to `w1`, for example, will be as following.

![](https://i.ibb.co/s2X9rkQ/backprop17.jpg)

We can find the update formula for the remaining weights w2, w3 and w4 in the same way. In summary, the update formulas for all weights will be as following:

![](https://i.ibb.co/6nY7nyr/backprop20.jpg)

We can rewrite the update formulas in matrices as following:

![](https://i.ibb.co/nkQwRZs/backprop19.jpg)

### Backward Pass

Using derived formulas we can find the **new weights**.

**Learning rate**: is a hyperparameter which means that we need to manually guess its value. 

∆ = 0.184 - 1 = -0.816

*Delta = prediction - actual*

a = 0.05 

*We choose a small number*


$\begin{bmatrix}w5 \\ w6\end{bmatrix} $ = $\begin{bmatrix}0.18 \\ 0.13\end{bmatrix} $ - 0.05(0.816) $\begin{bmatrix}0.71 \\ 0.43\end{bmatrix} $ = $\begin{bmatrix}0.18 \\ 0.13\end{bmatrix} $ - $\begin{bmatrix}-0.029 \\ -0.017\end{bmatrix} $ = $\begin{bmatrix}0.21 \\ 0.15\end{bmatrix} $

$\begin{bmatrix}w1 & w3 \\ w2 & w4\end{bmatrix}$ = $\begin{bmatrix}0.22 & 0.05 \\ 0.09 & 0.11\end{bmatrix}$ - 0.05(-0.816)$\begin{bmatrix}2\\ 3\end{bmatrix} $.$\begin{bmatrix}0.18\ 0.13\end{bmatrix} $ = $\begin{bmatrix}0.22 & 0.05 \\ 0.09 & 0.11\end{bmatrix}$ - $\begin{bmatrix}-0.015 & -0.01 \\ -0.022 & -0.016\end{bmatrix}$ = $\begin{bmatrix}0.23 & 0.06 \\ 0.11 & 0.13\end{bmatrix}$


Now, using the **new weights** we will repeat the forward pass:

![](https://i.ibb.co/N6N2qCr/newbackprop.jpg)

$\begin{bmatrix}2 \ 3\end{bmatrix} $. $\begin{bmatrix}0.23 & 0.06 \\ 0.11 & 0.13\end{bmatrix}$ = $\begin{bmatrix}0.79\ 0.51\end{bmatrix} $

$\begin{bmatrix}0.79\ 0.51\end{bmatrix} $.$\begin{bmatrix}0.21\\ 0.15\end{bmatrix} $ = [0.242]

2 x 0.23 + 3 x 0.11 = 0.79 

2 x 0.06+ 3 x 0.13 = 0.51

0.79 x 0.21 + 0.51 x 0.15 = 0.242

We can notice that the prediction **0.242** is a little bit closer to actual output than the previously predicted one 0.191. We can repeat the same process of backward and forward pass until error is close or equal to zero.


# 2B Assignment

The following section will implement the above backpropagation calculation in Python.

To start with we had:

inputs = [2, 3]

output = [1]

weights = [w1, w2, w3, w4, w5, w6] = [0.22, 0.09, 0.05, 0.11, 0.18, 0.13]

We will import numpy library to do the matrix opration to do the following:
* Forward pass to predict the output
* Calculate the difference between prediction and actual output. 
* Update the weights considering the derived formula from assignment 2A part
* Forward pass again to predict output with updated weights

### Forward pass

In [0]:
# Importing required library for matrix operation
import numpy as np

# initializing the weights with random values same as considered in assignment 2A
w1 = 0.22
w2 = 0.09
w3 = 0.05
w4 = 0.11
w5 = 0.18
w6 = 0.13

# define input array with corresponding values
inputs = np.array([[2,3]], dtype=np.float32)

# define array of weights for inputs to hidden layer
weights_1 = np.array([[w1, w3], [w2, w4]], dtype=np.float32)

# calculate the values at the nodes in hidden layer
h_val = inputs.dot(weights_1)

# define array for weights for inputs to output layer
weights_2 = np.array([w5, w6], dtype=np.float32)

# calculate the output value
output = h_val.dot(weights_2)

print('output:', output)
print(output.dtype)

output: [0.18370001]
float32


**Comparison:**

In the manual calculation we had rounded the calculated output value to three decimal places. Hence, the output value, as mentioned in assignment 2A is 0.184.

However, in the code, we are not rounding any number. And as per the set data type, all the matrix elements including the output value here are considered with 32 decimal places.

### Calculate Delta

In [0]:
# calculate the difference between the predicted value and actual output value
delta = output[0] - 1

# set the value of learning rate
a = 0.05

print('delta: ', delta)

delta:  -0.8162999898195267


### Backward Pass

In [0]:
# considering the values of delta and learning rate, calculate the new values of weights of inputs to the output layer

weights_2_u = np.array([0.18, 0.13], dtype=np.float32) - (delta * a) * h_val
print('Updated weights for inputs to output layer: \n', weights_2_u, '\n')

# similarly calculate the new values of weights of inputs to the hidden layer
weights_1_u = weights_1 - np.multiply(((delta * a) * inputs.T), weights_2.T)
print('Updated weights for inputs to hidden layer: \n', weights_1_u)

Updated weights for inputs to output layer: 
 [[0.20897865 0.14755045]] 

Updated weights for inputs to hidden layer: 
 [[0.2346934  0.0606119 ]
 [0.1120401  0.12591785]]


**Comparison**

The updated weights from the manual calculation (in assignment 2A):

$\begin{bmatrix}w1 \\ w2 \end{bmatrix}$ = $\begin{bmatrix}0.21 \\ 0.15\end{bmatrix}$

$\begin{bmatrix}w1 & w3 \\ w2 & w4\end{bmatrix}$ = $\begin{bmatrix}0.23 & 0.06 \\ 0.11 & 0.13\end{bmatrix}$

### Forward pass again

In [0]:
# calculate the updated values at the hidden nodes
h_val_u = inputs.dot(weights_1_u)

# calculate the output value with updated weights
output_u = h_val_u.dot(weights_2_u.T)

print("Updated values at hidden nodes: \n", h_val_u)

print("\nUpdated output value: ", output_u)

Updated values at hidden nodes: 
 [[0.80550706 0.49897736]]

Updated output value:  [[0.24195811]]


**Comparison**

Point to note here is that the calculated values at the nodes are slightly different from the corresponding ones from manual calculation because of the difference in decimal places considered.


### Outputs from manual calculation and Python implementation

Output from manual calculation (assignment 2A): **0.242**

Output from python implementation (assignment 2B): **0.24195811**

Interestingly, both the values are pretty close. Of course, the value from python implementation is more accurate, given that all the numbers were considered with 32 decimal places.